IanGRap/Squish
# Treasure Grab Part 2

Contributor: Coulter

Goal: The PR has two main purposes. The first is to connect the treasure grab scene to the win screen, ensuring that the scene transition happens properly. The second main goal is to finish the treasure grab manager script and ensure all mechanics work properly. Later on, the PR was expanded to also restructure the point system to make it more easily applicable to all game modes.

Files Changed:
- ActivePlayers.cs: This one is fairly important. Players now have two separate scores - points and wins. Points are for any game mode where points are earned during matches, and wins are the number of matches won by a player. For example, In Treasure Grab, breaking treasure blocks earns points, and the first player to reach 10 points gets a win for that match.
- GameManager.cs: moved code that controls win scene transition out of UIMatchManager.cs and into GameManager.cs. This script also stores the variable for the number of match wins needed to win overall.
- TreasureGrabModeManager.cs: Contains the functions for giving players points from treasure blocks, and taking away points for death. When the needed number of points is reached, calls the MatchEndManager.cs functions. Debug line also added to track when players gain or lose points, since the UI for that will be in the next PR.
- [Treasure Grab Match Manager].prefab: This prefab is where the variables for number of points and number of wins can be changed. Ideally (for now), the Treasure Grab mode would just be a single match, so whomever gets one win first, would win overall.
- UIMatchManager.cs: Moved win scene transition code out of this script and into GameManager.cs. A future additional fix to consider would be doing the same for ends of matches.

How to test:
- Start up the Treasure Grab Test Scene.
- Look at the values for required wins and points in the treasure grab match manager.
- Based on that, start the scene and play. Once the correct number of points is acquired, the end of the match should trigger. Once that happens, the player with the winning number of points should receive the win. The needed number of wins will trigger the win scene transition. This should essentially all work the same as in the normal survival mode, the main difference being the addition of points that are acquired and lost during play.
- Observe the win scene. Make sure that the players are placed on the podium in the correct order.
- Go back and change the required point and required win variables to ensure that matches and games end as they should each time.
- Go to the Match scene and make sure the survival mode still works as it always has. I'm pretty certain my changes haven't messed up survival mode's functionality.

Notes: This should be the last larger PR for Treasure Grab. Future ones should all be significantly smaller. The next one is UI to track points during play, and after that will be things like bringing in the treasure chest model that Wesley made, and creating a visual of coins transferring to the player when a chest breaks. Future PRs will also be for tweaking gameplay after the finished mode gets some more feedback.
